File: 20-ValidParentheses/20-ValidParentheses.java
// Last updated: 20/07/2026, 18:40:45
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>(); 
        for (char c : s.toCharArray()) { 
            if (c == '(') 
                stack.push(')'); 
            else if (c == '{') 
                stack.push('}');
            else if (c == '[')
                stack.push(']'); 
            else if (stack.isEmpty() || stack.pop() != c) 
               
                
                return false;
        }
        
           return stack.isEmpty();
    }
}

File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 20/07/2026, 18:40:37
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        # mapping={")":"(","}":"{","]":"["}
        for ch in s:
            if ch=="(":
                stack.append(")")
            elif ch=="[":
                stack.append("]")
            elif ch=="{":
                stack.append("}")
            else:
                if not stack or stack.pop()!=ch:
                    return False

        return len(stack)==0
            
        #     if char in mapping.values():
        #         stack.append(char)

        #     else:
        #         if not stack or stack[-1]!=mapping[char]:
        #             return False
        #         stack.pop()
        # return not stack















    #     stack = []
    #     mapping = {')': '(', '}': '{', ']': '['}

    #     for char in s:
    #         if char in mapping.values():  
    #             stack.append(char)
    #         else:  
    #             if not stack or stack[-1] != mapping[char]:
    #                 return False
    #             stack.pop()

    #     return not stack
